token_contracts/baln/balance.py

```diff
@@ -522,6 +532,138 @@ def burnFrom(self, _account: Address, _amount: int) -> None:
+        paying rewards, and paying dividends.
+        """
+        return (self.now() - self._time_offset.get()) // DAY_TO_MICROSECOND
+
+    # ----------------------------------------------------------
+    # Snapshots
+    # ----------------------------------------------------------
+
+    def _update_baln_staked_snapshot(self, _account: Address) -> None:
+        current_id = self.getDay()
+        current_value = self.detailsBalanceOf(_account)['Staked balance']
+        length = self._stake_snapshot[_account]['length'][0]
+
+        if length == 0:
+            self._stake_snapshot[_account]['amounts'][length] = current_value
+            self._stake_snapshot[_account]['length'][0] += 1
+            return
+        else:
+            last_snapshot_id = self._stake_snapshot[_account]['ids'][length - 1]
+
+        if last_snapshot_id < current_id:
+            self._stake_snapshot[_account]['ids'][length] = current_id
+            self._stake_snapshot[_account]['amounts'][length] = current_value
+            self._stake_snapshot[_account]['length'][0] += 1
+        else:
+            self._stake_snapshot[_account]['values'][length - 1] = current_value
+
+    def _update_total_staked_snapshot(self, _account: Address) -> None:
+        current_id = self.getDay()
+        current_value = self.detailsBalanceOf(_account)['Staked balance']
+        length = self._total_stake_snapshot[current_id]['length'][0]
+
+        if length == 0:
+            self._total_stake_snapshot[current_id]['amounts'][length] = current_value
+            self._total_stake_snapshot[current_id]['length'][0] += 1
+            return
+        else:
+            last_snapshot_id = self._total_stake_snapshot[current_id]['ids'][length - 1]
+
+        if last_snapshot_id < current_id:
+            self._total_stake_snapshot[current_id]['ids'][length] = current_id
+            self._total_stake_snapshot[current_id]['amounts'][length] = current_value
+            self._total_stake_snapshot[current_id]['length'][0] += 1
+        else:
+            self._total_stake_snapshot[current_id]['values'][length - 1] = current_value
+
+    @external(readonly=True)
+    def stakedBalanceOfAt(self, _account: Address, _day: int) -> int:
+        if _day < 0:
+            revert(f'{TAG}: snapshot id is equal to or greater then Zero')
+        low = 0
+        high = self._stake_snapshot[_account]['length'][0]
+
+        while low < high:
+            mid = (low + high) // 2
+            if self._stake_snapshot[_account]['ids'][mid] > _day:
+                high = mid
+            else:
+                low = mid + 1
+        if self._stake_snapshot[_account]['ids'][0] == _day:
+            return self._stake_snapshot[_account]['amounts'][0]
+        elif low == 0:
+            return 0
+        else:
+            return self._stake_snapshot[_account]['amounts'][low - 1]
+
+    @external(readonly=True)
+    def totalStakedBalanceOfAt(self, _day: int) -> int:
+        if _day < 0:
```

Comment:
Is zero value allowed? What would be the value for zeroth day ?

Reply:
if the day is 0, and value is not set on loading the value would be 0.